rs/replicated_state/src/canister_state/queues.rs

```diff
@@ -1259,23 +1404,50 @@ fn queue_front_not_stale(
     Ok(())
 }
 
-/// Collects the `CallbackIds` of all the responses enqueued in input queues.
-/// Returns an error if there are duplicate `CallbackIds` among the responses.
+/// Collects the `CallbackIds` of all responses and shed responses enqueued in
+/// input queues.
+///
+/// Returns an error if there are duplicate `CallbackIds` among the responses;
+/// or if not all inbound responses or shed responses are enqueued.
 ///
 /// Time complexity: `O(n * log(n))`.
 fn callbacks_with_enqueued_response(
-    _canister_queues: &BTreeMap<CanisterId, (CanisterQueue, CanisterQueue)>,
+    canister_queues: &BTreeMap<CanisterId, (CanisterQueue, CanisterQueue)>,
     pool: &MessagePool,
+    shed_responses: &BTreeMap<message_pool::Id, CallbackId>,
 ) -> Result<BTreeSet<CallbackId>, String> {
-    let mut callbacks = BTreeSet::new();
-    let duplicates: Vec<_> = pool
-        .inbound_response_callbacks()
-        .filter(|callback_id| !callbacks.insert(*callback_id))
-        .collect();
-    if !duplicates.is_empty() {
+    let callbacks_vec = canister_queues
```

Comment:
Technically, yes. Because we were previously directly looking at the pool contents and now we look them up by iterating over the queues (and comparing their respective lengths, at the very bottom).

But we are merely promoting a soft invariant that we were already testing at loading time (matching set of callbacks and no duplicates) to a hard invariant. And we've never seen that particular critical error.

So to the extent that it is ever safe to ever promote a soft invariant to a hard invariant, this is safe.